cmd/opentelemetry-prometheus-sidecar/main.go

```diff
@@ -237,52 +295,28 @@ func Main() bool {
 	level.Debug(logger).Log("msg", "entering run state")
 	healthChecker.SetRunning()
 
-	// Run two inter-depdendent components:
-	// (1) Prometheus reader
-	// (2) Queue manager
-	// TODO: Replace this with x/sync/errgroup
-	var g run.Group
-	{
-		g.Add(
-			func() error {
-				level.Info(logger).Log("msg", "starting Prometheus reader")
-				err = prometheusReader.Run(ctx, startOffset)
-				return err
-			},
-			func(err error) {
-				// Prometheus reader needs to be stopped before closing the TSDB
-				// so that it doesn't try to write samples to a closed storage.
-				// See the use of `stopCh` below to explain how this works.
-				level.Info(logger).Log("msg", "stopping Prometheus reader")
-				cancelMain()
-			},
-		)
-	}
-	{
-		stopCh := make(chan struct{})
-		g.Add(
-			func() error {
-				if err := queueManager.Start(); err != nil {
-					return err
+	g := getRunGroup(ctx, cancelMain, logger, startOffset, prometheusReader, queueManager)
+
+	for {
+		if err := g.Run(); err != nil {
+			if err == tail.ErrRestartReader {
```

Comment:
Q: I was expecting this error to be wrapped in various ways and thought we'd probably still need to use a string-match predicate here? (If that's not the case, excellent. It's sometimes tricky to follow errors through a run-group like this.)

Reply:
Ah you're right, I just remembered it gets wrapped in a corruption error. Will update